### ISSUE TITLE
Modify 'su', 'bs' definition to report both True and False boolean values

### ISSUE THREAD

**Issue by cksqa** [cta-wave/common-media-client-data#109]
@wilaw , as per discussion on video-dev #cmcd slack:

Christopher Kulbakas
  [12:08 PM](https://video-dev.slack.com/archives/C01SSSE4ZB5/p1694189294550709)
[@Will Law](https://video-dev.slack.com/team/U2NV3JDEX)
 , all CMCD keys are optional for a player to send , and ‘su’ flag , if opted by player to send, is not sent if ‘su’ is false. This leads to the ambiguity where lack of ‘su’ keys in CDN log may mean two different things: 1) ‘su’ is false ; 2) ‘su’ is opted out. Am I understanding this correctly?

Will Law
  [9:01 AM](https://video-dev.slack.com/archives/C01SSSE4ZB5/p1694437275762449)
Correct. So the absence of 'su' should not be interpreted as meaning anything. The thinking was that a  CDN would presumably only implement some sort of 'su' optimization in the positive case that 'su' was present. But you raise a good point - what if the CDN is intentionally de-optimizing content not flagged with 'su'? They would then hurt playback for a player that is not participating in the 'su' scheme. I feel that this case is simple enough to avoid by a) never de-prioritize non-su traffic and b) if you do, then have a business agreement or delivery property configuration that allows you to turn it off by device type/contentID etc.

Christopher Kulbakas
Hi 
[@Will Law](https://video-dev.slack.com/team/U2NV3JDEX)
, I am thinking about the monitoring use case, although the CDN optimization use case is also a great example. In circumstances where it is not the case that all players have the BS, SU flags enabled, we are at risk of not getting a complete or accurate picture of streaming "health". Also, because the configuration is controlled at the player, it may be difficult to enforce and/or track which players have the BS, SU flags enabled, especially if the player is controlled by a third-party. There is also the possibility that entities controlling player configurations may choose to carry out CMCD-based monitoring by way of sampling where they only enable a (changing-over-time) subset of players to send the CMCD BS & SU. This may render the BS and SU flag to be "untrusted" indicators, which could be a missed opportunity since startup time and buffer starvation obtained directly from the player are key indicators of streaming health. Are we over-engineering here for the sake of brevity at the cost of more complex configuration and management? (edited)

**Comment by gwendalsimon** [comment]
I can confirm that our Analytics team was quite unhappy with the 'bs' flag happening from time to time. 

The first request from a given client is a good indicator of whether a key is present or absent. In our observations, these keys are then consistently reported during the session... except the 'bs' key. We are tempted to simply ignore the 'bs' key (as many other keys we are not interested in) but the 'per-client rebuffering rate' metrics is popular among administrators. So, we considered adding the 'bs=false' ourselves in all CMCD logs, just to be sure in case 'bs' is reported, but indeed we should take the 'per-client rebuffering rate' metrics with a grain of salt because every client at 0 can be either a client wo/ problem or a client with a CMCD 'bs' reporting problem.

Long story short, I agree that it is "over-engineering for the sake of brevity at the cost of more complex configuration and management".

**Comment by cksqa** [comment]
Thank you for the discussion thus far @wilaw , @gwendalsimon -- please see below some additional thoughts -- please note my perspective on this: I would like to deliver a global, real-time, "trusted", player 'start up state' and/or 'buffer starvation state' dashboard to my NOC and Operational teams based on CMCD 'su','bs' keys:

>> IF CMCD is being received at all, AND su and bs are absent, THEN we can infer that the player is not in startup mode or rebuffering.

I don't think this is a valid statement for case 2) below, please correct me if I'm mistaken.

1) Player implementations not implementing 'bs','su' is one scenario. 2) Another scenario is that the player does implement 'bs','su' however, during runtime, the player is configured not to send 'bs','su' (i.e. subset of CMCD keys being reported by player is modified / dynamic across player population over time, e.g. sample-based monitoring, e.g. root-cause-analysis deep dive monitoring).

Combining 1) and 2) with situations where a CDN is servicing many different OTT services, each OTT service being controlled by different parties, and hence with different 'su','bs' implementation or configuration policies. This can lead to complexity with respect to tracking state of third-party configuration policies or even making configuration tracking impossible if configuration is dynamically turned on and off. This can lead to confusion as to how exactly to interpret the 'su',bs' flags at scale, i.e. we can't trust the observed 'bs' % rates at aggregate.

If we had a continuous TRUE/FALSE stream of 'su','bs' status, we would then know with confidence 1) what % of CDN-serviced population has 'su','bs' enabled ; 2) for the players with 'su','bs' enabled, what are the 'su','bs' % error rates.

>> "The reason we made su and bs value-free was because they occur so infrequently (<0.1% of requests?) so reporting su=false and bs=false for the majority of requests seems like a lot of overhead for little information transfer."

I would argue that this is precisely why we need consistent, accurate and complete reporting. Many difficult-to-diagnose issues occur infrequently (e.g. <0.1%), inconsistently and sporadically which makes diagnosing and troubleshooting them that much harder. In my opinion, showing a trusted, continuous 'everything is OK' signal for the >=99.9% is just as valuable as showing a 'NOT OK' signal for the <0.1%. Such a "trusted" continuous 'bs' True/False stream could become a key input for NOC/OPS situational awareness at scale.

I would also argue that other CMCD keys may also change infrequently, so why report them with every request? For example, if Top bitrate is implemented as reporting the BW value from the HLS manifest, then this may not change for the entire streaming session, and across all streaming sessions for a particular asset, for months or years.

**Comment by wilaw** [comment]
Reporting su=false and bs=false on every request gives you the certainty that the player is following the convention, but it comes at the expense of a lot of overhead. We repeat the same information and 99.9% of the time it is not needed. 

We could achieve the same certainty for the analytics system if the player _gave an assurance once per playback session_ that it would honor su and bs . For example, if the first time a player uses a SID (usually on a playlist or manifest request), it could attach new keys 'supportsSU=true' and 'supportsBS=true' . Since the SID is repeated across every request in that session, we can use it associate persistent state, such as whether that session is honoring su/bs, without having to continually repeat that information. @cksqa  - could that be an approach that would meet your analytics needs? I guess it boils down to do we need to be stateful or stateless in how we transmit the data?

**Comment by ZmGorynych** [comment]
We can achieve the above by specifying the keys to be reported in the manifest. This option is defined in DASH (AMD2 to 5th edition) along with the initial CMCD support.

**Comment by cksqa** [comment]
@wilaw - re "if the player gave an assurance once per playback session that it would honor su and bs" -- yes, I think this could be a good compromise -- I believe that when doing analytics processing, the analytics system could check for this condition when parsing and aggregating results on a per session basis -- @gwendalsimon , any thoughts from your end?

**Comment by gwendalsimon** [comment]
Yes, sorry, I mentioned IRL to Will that it is a good compromise but forgot to write it down here!

**Comment by nicolaslevy** [comment]
Regarding this, I see several options to address the issue (I will try to summarize some ideas and add new ones):

1.Mandatory True/False for 'bs' and 'su' Flags:
**Benefits**: Simplifies implementations, removes ambiguity about the meaning of flags.
**Drawbacks**: Results in repetitive data transmission.

2. Always Present Flag Control with CMSD Modification:
**Benefits**: Ensures consistent monitoring.
**Drawbacks**: High dependency between CMSD and CMCD for monitoring goals. Currently, CMSD isn't sufficiently developed across various players.

3.Binary Flag Mapping ('Player State'):
**Benefits**: Efficient data usage. Additional flags can be added by expanding the integer size.
**Drawbacks**: Not immediately clear to visually interpret player state; older fields may need to be deprecated in future versions.

4.Utilizing 'DASH (AMD2 to 5th edition)' as Suggested by Alex Giladi:
**Benefits**: There is a standard we can use
**Drawbacks**: Dependency on the manifest which might require manipulation or modification for desired behavior, potentially causing implementation friction.
Each option offers distinct approaches to enhance CMCD's effectiveness in managing streaming health, considering both the CDN optimization and monitoring use cases. The choice largely depends on the balance between implementation simplicity, data efficiency, and monitoring accuracy.

**Comment by PCaponetti** [comment]
Will - decision to make, keep things stateless or stateful
Alex - wanting to keep things stateless, +1 from Chris
Paul - rely on logging aggregation where possible, but needing signal in the request for real-time decisioning in the data plane
David - other use cases where having state would be valuable
Chris K - +1 for Will's proposal for session level
Paul - how many other keys will need this type of mechanism?
Nicholas - worrying about the tristate
Paul - could also add MUST for the true state
Jim H - only issue for su and bs was formatting the json required a lot of extra work because it's hard to represent it as json text, +1 for bs=true as opposed to bs being present and adding MUST language
Chris L - using structured field values, so put them in if true and leave out if false. 
Will - just added an issue for integer values for what were boolean values, which begs the question of whether or not we need to be backward compatible with v2
Glenn - looking at buffer length, and possibly buffer length 0 could signal what we need?
Will - the correlation of rebuffering and buffer length == 0 is not robust
Chris L - not maintaining backwards compatibility is a cost that needs to be evaluated. the cost may be in terms of CMCD v2 penetration
Chris L - the main question: do we care if a player is experiencing starvation and not telling us?
Will - bs is not a state of the player, but an event, and think it should stay that way and enrich it with the duration of that event
Chris L - if we report the duration of the rebuffer, the value of 0 is an indication of success?
Will - buffering is a very minimal hardship for the player to have to report

Wrap up was to suggest MUST language in the spec so that not reporting these values is the same as a false value for the boolean.

Leaving this open until next session for soak time on the suggestion.

**Comment by littlespex** [comment]
Just a reminder, using [structured field values](https://www.rfc-editor.org/rfc/rfc8941.html#name-serializing-a-boolean) `false` would be encoded as `?0`. `bs=?0` and `su=?0`.

**Comment by cksqa** [comment]
To bring this back to the original intent of this request: a foundational aspect of any analytics system is correctness and validity of, and consequently, operational trust in, the collected data:

The current definition introduces an ambiguity ('su', 'bs' have 2 possible states and we don't know which one is the actual one: 1) not implemented; 2) implemented and false). This can lead to situations that make the monitoring conclusions invalid (ex. we don't see any 'bs' flags hence we can conclude that no buffer starvation is occurring yet buffer starvation is occurring but the players are not reporting it).

In my opinion, this approach is not practical because relying on checking and keeping tabs on which players have which flags implemented is too much of a burden on analytics systems, may be even impossible if OTTs don't share player implementation state with CDNs (or any other CMCD data collectors, i.e. Glenn's suggestion for a generic CMCD data export feature) or the selective enablement / disablement of the flags being reported as part of A/B testing, etc.

Summarising my take based on the excellent discussion today:

1) Making the implementation of bs, su a MUST is one way to address this, however, we can run into issues with implementation penetration (as per Chris L point) and we are now on a slippery slope, i.e. prescribing implementation.

2) Making the reporting of each of the tri-states (false, true, not implemented) with every object request for bs, su state a MUST is another way, however, it does increase on-the-wire BW usage due to the continuous reporting of su, bs state with every request, however changing the field type to integer to indicate the tri-state (0,1,2) could reduce BW usage.

3) Making the reporting on a per session level whether bs, su is enabled a MUST is another way to address this, but requires to keep session state, however, we already do keep session state with fields like content ID, playback rate, content format, so what makes bs, su different in that respect?
